### DISCUSSION TITLE
how to get MySQL warning count after upsert?

### DISCUSSION THREAD

**Discussion by HarryCheck** [sqlalchemy/sqlalchemy#8915] [Usage Questions] [answered]
### Describe the use case

Accroding to 
[https://mysqlclient.readthedocs.io/user_guide.html](https://mysqlclient.readthedocs.io/user_guide.html)
The mysqldb api provide a method to retrieve the warning count of the db.

### Databases / Backends / Drivers targeted

It is for mysql database.

### Example Use

These is particularly usful when using `insert ignore .....`  to get how many duplicates key you encounter.

Example use:
`
result = engine.execute('insert ignore .....') 
`
`
duplicate_count = result.warning_count()
`

**Comment by zzzeek** [comment, marked as answer]
hi -

this is a non-standard feature that is specific to mysqlclient, and not any of the other drivers.  you would use the [raw connection](https://docs.sqlalchemy.org/en/14/core/connections.html#working-with-the-dbapi-cursor-directly) to get this

```py
with engine.connect() as conn:
    conn.execute(text("insert ..."))
    warning_count = conn.connection.warning_count()
```

more DBAPI agnostic, just use SHOW WARNINGS directly, this would be the better option:

```py
with engine.connect() as conn:
    conn.execute(text("insert ..."))
    warning_count = conn.exec_driver_sql("SELECT @@warning_count").scalar()
```

that's what warning_count() does anyway.
